lib/LaTeXML/MathParser.pm

```diff
@@ -680,19 +693,85 @@ sub parse_single {
+        my $font_text = '';
+        foreach my $attr (keys %pending_declaration) {
+          if ($document->canHaveAttribute($qname, $attr)) {
+            $font_text .= $pending_declaration{$attr}{value}; } }
+        if ($font_text) {
+          $font_text = join("-", sort(split(/\s+/, $font_text)));
+          $lexeme = $font_text . "-" . $lexeme; } } } }
+  if (my $role = $self->getGrammaticalRole($node)) {
+    if ($role ne 'UNKNOWN') {
+      $lexeme = $role . ":" . $lexeme; } }
+
+  $lexeme =~ s/\s//g;
+  return $lexeme;
+}
+
+sub node_to_lexeme_full {
+  my ($self, $unrealized_node) = @_;
+  my $node = realizeXMNode($unrealized_node);
+  my $tag  = getQName($node);
+  if ($tag eq 'ltx:XMHint') { return ""; }    # just skip XMHints, they don't contain lexemes
+  my $role = p_getAttribute($node, 'role');
+  if (($tag =~ /^ltx:XM(Tok|Text)$/) || ($role && ($tag !~ 'ltx:XM(Dual|App|Arg|Wrap|ath)'))) {
+# Elements that directly represent a lexeme, or intended operation with a syntactic role (such as a postscript),
+# can proceed to building the lexeme from the leaf node.
+    return $self->node_to_lexeme($node);
+  } else {
+# Elements that do not have a role and are intermediate "may" need an argument wrapper, so that arguments
+# remain unambiguous. For instance a `\frac{a}{b}` has clear argument structure to be preserved.
+    my ($mark_start, $mark_end) = ('', '');
+    if ($tag ne 'ltx:XMath') {
+      if ($role) {
+        $mark_start = "$role:start ";
+        $mark_end   = "$role:end";
+      } else {
+        if ($tag eq 'ltx:XMArg') {
+          $mark_start = "ARG:start ";
+          $mark_end   = "ARG:end";
+        }
+      }
+    }
+    my $lexemes = $mark_start;
+    if ($tag eq 'ltx:XMDual') {
+      $lexemes .= $self->node_to_lexeme_full($LaTeXML::MathParser::DOCUMENT->getSecondChildElement($node));
+    } else {
+      my @child_nodes = element_nodes($node);
+      # skip through single child wrappers (don't serialize)
+      while (scalar(@child_nodes) == 1 && getQName($child_nodes[0]) =~ /^ltx:XM(Arg|Wrap)$/) {
+        @child_nodes = element_nodes($child_nodes[0])
+      }
+      foreach my $child (@child_nodes) {
+        my $child_lexeme = $self->node_to_lexeme_full($child);
+        $lexemes .= $child_lexeme . ' ' if $child_lexeme;
+      }
+    }
+    $lexemes .= $mark_end;
+    return $lexemes;
+  }
+}
+
 sub parse_internal {
   my ($self, $rule, @nodes) = @_;
   #------------
   # Generate a textual token for each node; The parser operates on this encoded string.
   local $LaTeXML::MathParser::LEXEMES = {};
   my $i       = 0;
   my $lexemes = '';
+
   foreach my $node (@nodes) {
-    my $role = $self->getGrammaticalRole($node);
-    my $text = getTokenMeaning($node);
-    $text = 'Unknown' unless defined $text;
-    my $lexeme = $role . ":" . $text . ":" . ++$i;
-    $lexeme =~ s/\s//g;
+    my $lexeme = $self->node_to_lexeme($node) . ":" . ++$i;
```

Comment:
ah, I forgot I was reusing the method in the internal parse, I need to be a little more careful here - my font changes broke a test I'm afraid.